tracepoint_decode/src/charconv.rs

```diff
@@ -0,0 +1,265 @@
+    }
+}
+
+// **** write_latin1_to
+
+/// Writes a Latin-1-encoded string to a filter.
+pub fn write_latin1_to<F: filters::Filter>(bytes: &[u8], filter: &mut F) -> fmt::Result {
+    let len = bytes.len();
+    let mut written_pos = 0;
+    for pos in 0..len {
+        let b = bytes[pos];
+        if b <= 0x7F {
+            // ASCII. Continue.
+            continue;
+        }
+
+        // bytes[pos..] is not ASCII.
+
+        // Flush the ASCII, if any.
+        if written_pos < pos {
+            // Validated: substring contains only ASCII.
+            filter.write_str(str_from_validated_utf8(&bytes[written_pos..pos]))?;
+        }
+
+        filter.write_non_ascii(b as char)?;
+        written_pos = pos + 1;
+    }
+
+    // Write any remaining ASCII. (Common case: the entire string.)
+    return if written_pos < len {
+        // Validated: substring contains only ASCII.
+        filter.write_str(str_from_validated_utf8(&bytes[written_pos..]))
+    } else {
+        Ok(())
+    };
+}
+
+// **** write_utf8_with_latin1_fallback_to
+
+/// Writes a UTF-8-encoded string to a filter. If the string contains any invalid UTF-8 sequences,
+/// the sequences are treated as Latin-1.
+pub fn write_utf8_with_latin1_fallback_to<F: filters::Filter>(
+    bytes: &[u8],
+    filter: &mut F,
+) -> fmt::Result {
+    let len = bytes.len();
+    let mut written_pos = 0;
+    let mut pos = 0;
+    while pos < len {
+        // If this is valid UTF-8, update pos and continue to next iteration.
+        // If this is not valid UTF-8, fall-through to the Latin1 case.
+        let b0 = bytes[pos];
+
+        if b0 <= 0x7F {
+            // 0x00..0x7F: Valid UTF-8. Continue.
+            pos += 1;
+            continue;
+        } else if b0 <= 0xBF {
+            // Invalid lead byte. Fall-through.
+        } else if b0 <= 0xDF {
+            if len - pos >= 2 {
+                let b1 = bytes[pos + 1];
+                if 0x80 == (b1 & 0xC0) {
+                    let ch = ((b0 & 0x1F) as u32) << 6 | ((b1 & 0x3F) as u32);
+                    if 0x80 <= ch {
+                        // Valid 2-byte UTF-8. Continue.
+                        pos += 2;
+                        continue;
+                    }
+                }
+            }
+        } else if b0 <= 0xEF {
+            if len - pos >= 3 {
+                let b1 = bytes[pos + 1];
+                let b2 = bytes[pos + 2];
+                if 0x80 == (b1 & 0xC0) && 0x80 == (b2 & 0xC0) {
+                    let ch = ((b0 & 0x0F) as u32) << 12
+                        | ((b1 & 0x3F) as u32) << 6
+                        | ((b2 & 0x3F) as u32);
+                    if 0x800 <= ch && !(0xD800..=0xDFFF).contains(&ch) {
+                        // Valid 3-byte UTF-8. Continue.
+                        pos += 3;
+                        continue;
+                    }
+                }
+            }
+        } else if b0 <= 0xF4 {
+            #[allow(clippy::collapsible_if)]
```

Comment:
This should be viewed with immediate suspicion.

Why?

Reply:
Because "The symmetry seems helpful in understanding this code".